### PR TITLE
Playwright: QoL improvements to the authentication cookie mechanism

### LINKED ISSUE #57948: Playwright: refinements to auth cookie login method
### Details

The authentication cookie saving & login mechanism was implemented in https://github.com/Automattic/wp-calypso/pull/57569.

There are some tweaks that can be made to make it a better experience locally and on CI. See the checklist.

### Checklist

- [x] reduce the wait period to a shorter timeout.
- [x] use timestamp of cookie when last modified, and if too old (> 1 day) then skip the cookie attempt.

### Related

_No response_

### PR DESCRIPTION
**Pull request by worldomonation**
#### Changes proposed in this Pull Request

This PR proposes some quality of life improvements to the authentication cookie saving mechanism.

Key changes:
- check whether the cookie file is recent enough (3 days) in `globalSetup`.
- if the cookie file is recent, skip the authentication & save process.
- additional check in `BrowserManager.setLoginCookie` to check whether the cookie file found is recent.
- if the cookie file is not recent, go straight to the traditional username/password login branch.

Details:
These changes are primarily aimed at users running the test suites locally. In a CI environment, no change should be observed.

When running tests locally and wanting to save the authenticated cookies, the environment variable `SAVE_AUTH_COOKIES` is either prepended to the `yarn jest` call, or exported permanently to the environment.

Prior to this PR, subsequent invocations of `yarn jest` must take care to remove the `SAVE_AUTH_COOKIES` env var, or otherwise spend up to 10 seconds while the cookies are refreshed. This is tedious when quickly debugging or fixing something.

With this PR, the authentication cookie saving mechanism will only replace the cookie file if the cookie is determined to be stale. If the cookie files are considered recent enough, `globalSetup` is short-circuited.

Further checks in `BrowserManager.setLoginCookie` also serves to short-circuit the branch in `LoginPage.login` if a stale cookie is found. Compare this to current method where 15 seconds must elapse before the traditional login using username/password is attempted.

#### Testing instructions

- [x] tests execute as expected.

Related to #

### CODE REVIEW

File: packages/calypso-e2e/src/browser-manager.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/browser-manager.ts:216]: I'm using the file creation time as a stand-in for the time when the cookie was created, but I fully acknowledge that this is not a 100% foolproof method. Anyone wanting to trick this validation can easily do so by creating a new file and copying stale cookie contents into it.

With that said, there doesn't seem to be a 'good' or 'easy' way to check the expiry timestamp of a cookie. Since we have a cookie file we're working with, it's possible to iterate through the objects, looking for the `wordpress_logged_in` name then checking the expiry time, but that is:

1. tedious, and'
2. the cookies have expiry date of +3 years but don't actually last that long.

For 2, I am basing this from my understanding of PCYsg-5YE-p2.

File: packages/calypso-e2e/src/lib/pages/login-page.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/pages/login-page.ts:188]: Maybe this log isn't necessary, but I have included it so users know what's happening.

File: test/e2e/lib/jest/globalSetup.ts
Comment (worldomonation) [inline test/e2e/lib/jest/globalSetup.ts:62]: Either `BrowserContext.storageState` doesn't overwrite the file, or the overwritten file still has the `stat.birthTimeMS` of the old version of the file therefore is regarded as a 'stale' cookie.

File: packages/calypso-e2e/src/browser-manager.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/browser-manager.ts:216]: Hmm, it's looking like the cookie files are possibly stale after a day.

File: packages/calypso-e2e/src/browser-manager.ts
Comment (WunderBart) [inline packages/calypso-e2e/src/browser-manager.ts:209]: Let's use this format for inline comments only (`/* inline comment */`), and use `//` for multiline ones to stay consistent across the codebase. https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/#comments

File: packages/calypso-e2e/src/lib/pages/login-page.ts
Comment (WunderBart) [inline packages/calypso-e2e/src/lib/pages/login-page.ts:187]: Do we need to check the instance here? Will it retry the username/password-based authentication regardless of the error type?

File: test/e2e/lib/jest/globalSetup.ts
Comment (WunderBart) [inline test/e2e/lib/jest/globalSetup.ts:65]: Should we provide the same feedback [as above](https://github.com/Automattic/wp-calypso/pull/58267/files#diff-1868a0f5bc3c3ac7ed9f7b61df88fc1f401c8b9d97bb413570fdd9738bf31986R200)?

```js
throw new Error( `Cookie file ${ cookiePath } not found on disk.` );
```

**Review by WunderBart — APPROVED** [review]
Just a few small comments. Feel free to :shipit: once they're are addressed! 🚀

File: packages/calypso-e2e/src/browser-manager.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/browser-manager.ts:209]: Interesting, I thought it was the other way around - admittedly my last experience with Java-syntax languages (before A8C) was in university with Java where I used block comments for multi-line comments. 

I will change the comment format but I still think a block comment is better from a formatting perspective.

File: packages/calypso-e2e/src/lib/pages/login-page.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/pages/login-page.ts:187]: TypeScript tells me I need to check the `err` type in the catch block, so I've implemented it this way.
https://stackoverflow.com/a/68257472

File: test/e2e/lib/jest/globalSetup.ts
Comment (worldomonation) [reply test/e2e/lib/jest/globalSetup.ts:65]: The error is suppressed for a couple of reasons:

1. if an error is thrown in a `globalSetup` type file, it would block execution of the actual test suite.
2. user who has never run an end to end test locally will not have a cookie file created locally. If `unlink` is called on a path that does not exist, it will throw an exception, and for the reason cited in 1) we do not want to throw an error here.

Comment (worldomonation) [reply test/e2e/lib/jest/globalSetup.ts:65]: There's also the approach I applied here, which is to make the `globalSetup` as transparent as possible from a user perspective - they should not even realize what is happening and things should just 'work' in a transparent manner, preferably.

Hence my decision to only output the most essential logs only and never to throw an exception in the `globalSetup`.

File: packages/calypso-e2e/src/lib/pages/login-page.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/pages/login-page.ts:187]: For the second part of your question, yes, the normal username/password attempt will be made regardless of the error type.

File: test/e2e/lib/jest/globalSetup.ts
Comment (WunderBart) [reply test/e2e/lib/jest/globalSetup.ts:65]: Thanks for elaborating. IMO, it's difficult to say which part of the whole `try` block would throw so in the end it's hard to figure out what we're suppressing. If the whole block depends on the existence of the `cookiePath`, do you think something like this could be more transparent code-wise:

```js
import { access } from 'fs/promises';

// ...

let cookieFileExists = false;
try {
  await access( cookiePath );
  cookieFileExists = true;
} catch {
  // noop
}

if ( cookieFileExists ) {
  // Do things with the cookie file
} else {
  // Do things without the cookie file
}
```

File: packages/calypso-e2e/src/browser-manager.ts
Comment (WunderBart) [reply packages/calypso-e2e/src/browser-manager.ts:209]: For multiline comments, I believe the other popular standard (and the one used in WP *.php files) is the starred-block:

```js
/**
 * Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec lectus augue,
 * aliquam dapibus est eget, posuere ornare sem. Quisque nec cursus elit.
 * Maecenas egestas at nunc nec cursus. Curabitur efficitur id ex at pharetra.
 * Nam turpis risus, laoreet quis odio eget, mollis sodales justo. Morbi commodo
 * sem dignissim, porttitor ipsum ac, lobortis nulla.
 */
```

Comment (worldomonation) [reply packages/calypso-e2e/src/browser-manager.ts:209]: Oh duh, I totally forgot about the second star in the first line.

I forgot to push the changes here before I (overzealously) hit the merge button, so I've pushed the changes to https://github.com/Automattic/wp-calypso/pull/58728 and merged it together.

File: test/e2e/lib/jest/globalSetup.ts
Comment (worldomonation) [reply test/e2e/lib/jest/globalSetup.ts:65]: I think separating it out like that would allow for a more relevant console output if we deem it necessary, so I am generally on board with the proposal. I'll spin up a new task for now.

Comment (scinos) [reply test/e2e/lib/jest/globalSetup.ts:65]: Node recommends not using `access`, and instead of catching the exception later if the file doesn't exist (when calling `stats` in this case). The rationale an external process could have deleted the file between the call to `access` and the call to `stats`, so you need error control on `stats` anyway.

> Do not use fs.access() to check for the accessibility of a file before calling fs.open(), fs.readFile() or fs.writeFile(). Doing so introduces a race condition, since other processes may change the file's state between the two calls. Instead, user code should open/read/write the file directly and handle the error raised if the file is not accessible.

(From https://nodejs.org/api/fs.html#fsaccesspath-mode-callback)

**Pull request merged**
